Axvline & Axhline replotting on project load

joseph-torsney opened the pull request:
This pull request supports the saving and loading of projects that contain plots with `ax.axvline`s and `ax.axhline`s. 

This partially fixes issue #28985, as we are yet to investigate Mantid Line Markers.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** Aidy / ISIS
-->

**To test:**

<!-- Instructions for testing. -->
1. Run the following script
```
from mantid.simpleapi import *
import matplotlib.pyplot as plt

ws1 = CreateSampleWorkspace(OutputWorkspace="ws1")
fig, axes = plt.subplots(edgecolor='#ffffff', num='SANSLOQCan2D-1', subplot_kw={'projection': 'mantid'})
axes.plot(ws1, label='ws1', specNum=1)
axes.axhline(1, xmin=0, xmax=1)
axes.axvline(0, ymin=0, ymax=1)
plt.show()
```
2. With the plot still open, save the project.
3. Close the plot and open the project you just saved.
4. When the project loads, verify the same plot appears, with the axvline and axhline plotted correctly.

fixes _**Partially**_ #28985 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

joseph-torsney committed:
Single line replotted on project load

joseph-torsney committed:
Test line data is correctly added to dict

joseph-torsney committed:
Test axvline and axhline are replotted on load

StephenSmith25 commented:
The code in general looks good from what I've seen so far - although I have a few issues in testing, which could be local issues for me, so would be worth you testing as well:

- If I load an ordinary workspace e.g MUSR62260 plot a single spectrum and save the project It can't save the plot (it throws up a warning in the logger).
- Also when I load a project file I created on a nightly the legend is missing (as its throwing early when we create `ws_line_labels`)

I think this is an issue with getting the line label.  When a plot is created the line doesn't seem to have a label immedietly assigned to it. You can force it to have one by going to figure options, curves -> and changing the curve label and clicking apply. After this you'll be able to save the project (and the plot will be correctly saved).

StephenSmith25 commented on qt/python/mantidqt/project/plotsloader.py:
```diff
@@ -147,9 +148,25 @@ def restore_fig_axes(self, ax, dic):
+        # Flatten the creation args and get just the label from each dict.
+        # These labels are for lines associated with workspaces.
+        ws_line_labels = [
+            line_dict['label'] for sublist in self.creation_args for line_dict in sublist
```
I'm wondering if this issues are created here.

DanielMurphy22 commented:
Great to see this getting worked on! Just a small thing, 
> Partially Fixes #28985

to Github means *Fixes #28985* so it is going to close the issue! If this is not desired then change the description. You may prefer to raise a new issue describing the work left to do, linking to this issue!

joseph-torsney committed:
Axvline and axhline identified by xy data

Lines associated with workspace data did not necessarily have a label assigned to them, causing errors on plot save. We now identify axvlines and axhlines by the shape of their xy data, rather than labels.

joseph-torsney commented:
Closing this PR as we're going to have a discussion about how to handle saving 1D and 2D lines properly, to avoid having a jumble of code handling different special cases.
